python/ambassador/config/config.py

```diff
@@ -593,10 +593,18 @@ def safe_store(self, storage_name: str, resource: ACResource, allow_log: bool=Tr
         storage = self.config.setdefault(storage_name, {})
 
         if resource.name in storage:
-            # Oooops.
-            self.post_error("%s defines %s %s, which is already defined by %s" %
-                            (resource, resource.kind, resource.name, storage[resource.name].location),
-                            resource=resource)
+            if resource.namespace == storage[resource.name].get('namespace'):
+                # If the name and namespace, both match, then it's definitely an error.
+                # Oooops.
+                self.post_error("%s defines %s %s, which is already defined by %s" %
+                                (resource, resource.kind, resource.name, storage[resource.name].location),
+                                resource=resource)
+            else:
+                # Here, we deal with the case when multiple resources have the same name but they exist in different
+                # namespaces. Our current data structure to store resources is a flat string. Till we move to
+                # identifying resources with both, name and namespace, we change names of any subsequent resources with
+                # the same name here.
+                resource.name = f'{resource.name}.{resource.namespace}'
```

Comment:
Is there a reason why changing the resource name done in an `else` block? It currently feels like we handle an edge-case, yet I'm pretty sure all `resource.name` could include the namespace from the moment they are handled.
Following the logic here, it seems given resources: `{name: something, namespace: a}` and `{name: something, namespace: b}`, we would end up with 2 resources named: `something` and `something.b` where we don't associate the namespace to the first resource we put in storage.

Reply:
You're correct. This is what I tried to explain at https://github.com/datawire/ambassador/pull/2230#issue-366506982 that we need to identify all resources with both name and namespace, and that's what I tried to do in the first go - but it turned out to be a much more intrusive change for the following reasons -
- there were test failures everywhere
- there are parts of code that rely on the resource name coming from aconf.config instead of aconf.config['name'], so that approach broke it

So, yes, this is more or less handling of an edge case without breaking the status quo.

Reply:
We'll definitely need to handle the larger case, but, yeah, got it.